### PR TITLE
[mlir] Improve error message when number of operands and types differ

### PR DESCRIPTION
**Pull request by zero9178**
If using a variadic operand, the error message given if the number of types and operands do not match would be along the lines of:
```
3 operands present, but expected 2
```

This error message is confusing for multiple reasons, particular for beginners:
* If the intention is to have 3 operands, it does not point out why it expects 2. The user may actually just want to add a type to the type list
*  It reads as if a verifier error rather than a parser error, giving the impression the Op only supports 2 operands.

This PR attempts to improve the error message by first noting the issue ("number of operands and types mismatch") and mentioning how many operands and types it received.

### CODE REVIEW

File: mlir/include/mlir/IR/OpImplementation.h [hunk 1]
```diff
@@ -1602,14 +1602,33 @@ class OpAsmParser : public AsmParser {
...
+    size_t lesserQuantity = operandSize;
+    size_t higherQuantity = typeSize;
+    if (operandSize > typeSize) {
+      std::swap(lesserQuantity, higherQuantity);
+      std::swap(lesserQuantityText, higherQuantityText);
+    }
+
+    diag.attachNote() << "got " << higherQuantity << " " << higherQuantityText
+                      << " but only " << lesserQuantity << " "
+                      << lesserQuantityText;
```
Comment (kuhar) [inline mlir/include/mlir/IR/OpImplementation.h:1630]: What do you think about emitting something simpler like `got N operands and K types`?

Comment (zero9178) [reply mlir/include/mlir/IR/OpImplementation.h:1630]: I don't feel super strong about either options, so if you feel the current version is worse or not worth the extra implementation complexity I can gladly change it!

Comment (kuhar) [reply mlir/include/mlir/IR/OpImplementation.h:1630]: Yeah I feel like the current version is a little bit overcomplicated for little benefit over what I suggested above

Comment (River707) [reply mlir/include/mlir/IR/OpImplementation.h:1630]: I would also +1 keeping it simpler. I would also keep it as part of the main message and not a note.

**Review by kuhar — APPROVED** [review]
LGTM

**Pull request merged**